src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -58,152 +71,279 @@ makeSem = genFreer True
+-- issue #48).
+--
+-- Please, change this as soon as GHC provides some way of inspecting
+-- signatures, replacing code or generating haddock documentation in TH.
 
+------------------------------------------------------------------------------
 -- | Generates declarations and possibly signatures for functions to lift GADT
 -- constructors into 'Sem' actions.
 genFreer :: Bool -> Name -> Q [Dec]
-genFreer makeSigs tcName = do
-  -- The signatures for the generated definitions require FlexibleContexts.
-  isExtEnabled FlexibleContexts
-    >>= flip unless (fail "makeSem requires FlexibleContexts to be enabled")
-  hasTyFams <- isExtEnabled TypeFamilies
-
-  reify tcName >>= \case
-    TyConI (DataD _ _ _ _ cons _) -> do
-      sigs <- filter (const makeSigs) <$> mapM genSig cons
-      decs <- join <$> mapM genDecl cons
-      loc <- location
-
-      return $
-        [ TySynInstD ''DefiningModule
-            . TySynEqn [ConT tcName]
-            . LitT
-            . StrTyLit
-            $ loc_module loc
-        | hasTyFams
-        ] ++ sigs ++ decs
-
-    _ -> fail "makeSem expects a type constructor"
-
--- | Given the name of a GADT constructor, return the name of the corresponding
--- lifted function.
-getDeclName :: Name -> Name
-getDeclName = mkName . overFirst toLower . nameBase
- where
-  overFirst f (a : as) = f a : as
-  overFirst _ as       = as
-
--- | Builds a function definition of the form @x a b c = send $ X a b c@.
-genDecl :: Con -> Q [Dec]
-genDecl (ForallC _       _     con) = genDecl con
-genDecl (GadtC   [cName] tArgs _  ) = do
-  let fnName = getDeclName cName
-  let arity  = length tArgs - 1
-  dTypeVars <- forM [0 .. arity] $ const $ newName "a"
-  pure $
-    [PragmaD (InlineP fnName Inlinable ConLike AllPhases)
-    , FunD fnName . pure $ Clause
-        (VarP <$> dTypeVars)
-        (NormalB . AppE (VarE 'send) $ foldl
-          (\b -> AppE b . VarE)
-          (ConE cName)
-          dTypeVars
-        )
-        []
+genFreer should_mk_sigs type_name = do
+  checkExtensions [ScopedTypeVariables, FlexibleContexts]
+  dt_info    <- reifyDatatype type_name
+  cl_infos   <- traverse (mkCLInfo dt_info) $ datatypeCons dt_info
+  tyfams_on  <- isExtEnabled TypeFamilies
+  def_mod_fi <- sequence [ TySynInstD ''DefiningModule
+                             . TySynEqn [ConT $ datatypeName dt_info]
+                             . LitT
+                             . StrTyLit
+                             . loc_module
+                           <$> location
+                         | tyfams_on
+                         ]
+  decs       <- traverse (genDec should_mk_sigs) cl_infos
+
+  let sigs = [ genSig <$> cl_infos | should_mk_sigs ]
+
+  return $ join $ def_mod_fi : sigs ++ decs
+
+------------------------------------------------------------------------------
+-- | Generates signature for lifting function and type arguments to apply in
+-- it's body on effect's data constructor.
```

Comment:
The smallest of all nits: "its"

Reply:
Still waiting for English syntax checker :smile: